### ISSUE TITLE
Support refreshing the view

### ISSUE THREAD

**Issue by mathstuf** [nbedos/cistern#10]
It'd be nice to have an `r` (or similar) binding to refresh the call. It seems that once a pipeline is complete, no more requests are sent to update it. This is reasonable, however, if I retry a failed job, `citop` doesn't show it until I restart it.

**Comment by mathstuf** [comment]
If told that I want to watch `HEAD` or `master`, refreshing what commit this means should also be supported in some way.

**Comment by nbedos** [comment]
Yes, that's planned.

**Comment by simplelife2010** [comment]
It would also be nice if refreshing would also display new pipelines. Currently, if after running cistern a new pipeline starts, it is never displayed.

**Comment by mathstuf** [comment]
I've seen them show up if I leave it open overnight, but I sometimes have to refresh manually over the course of the day.

**Comment by simplelife2010** [comment]
I have a lot of pipelines starting over the day, they will not appear in the list. How do you refresh manually? I have not seen any key binding to do that.

I have to quit and restart.

**Comment by mathstuf** [comment]
You can press `r` to choose a ref. If it isn't different than before, it will "ignore" you, but you can swap between `master` and `HEAD` which should refresh things.

**Comment by simplelife2010** [comment]
Which version do you use? I don't seem to have this feature in my version cistern 0.2.0-5-ga124d04-darwin-amd64.

**Comment by mathstuf** [comment]
Ah, right. I was testing some pre-releases for fixes to other issues I opened. 0.2.0dev6-linux-amd64

**Comment by nbedos** [comment]
I made changes related to this issue on the following branch: https://github.com/nbedos/cistern/tree/feature/ref_navigation

You can now use the following keys:

* **r** or **F5** to refresh the view in order to list new pipelines or get updates about jobs that have been restarted
* **f** to follow the current git reference in case it changed (e.g. if you ran `cistern master` in a terminal and then added a commit to `master` from another terminal you can now press **f** to have `cistern` monitor the last commit instead of the second to last). 
* **g** to open a prompt to input the name of a git reference to monitor. For local repositories this comes with autocomplete to make our lives easier.

What do you think?

**Comment by simplelife2010** [comment]
I like.
Would it be possible to do the refresh automatically? Right now, status of already shown pipeline jobs including duration is automatically updated. For new pipelines I have to refresh manually, which works fine. But intuitively I would expect this the be done automatically.

The f key works fine.

The g key shows up a table with all the refs, but choosing one does not have an effect.

**Comment by nbedos** [comment]
Thanks for the feedback!

I pushed changes on the same branch that add the polling strategy used by `cistern` to the configuration file:

```toml
[providers.polling]
# cistern gets information from providers by sending requests to their API at increasing intervals.
# The duration between two consecutive requests for a given API resource starts at
# the value defined by the key "initial-interval" and exponentially increases until it reaches
# the duration "max-interval". Each time a requests returns new data (e.g. a new pipeline, or a
# pipeline that changed since the last request), the duration between requests is reset to the value
# of "initial-interval" and starts increasing again.

# Duration of the initial interval in seconds (integer, default: 10)
initial-interval = 10

# Duration of the maximal interval in seconds (integer, default: 120)
max-interval = 120

# Continue polling once the duration "max-interval" is reached. The duration between two requests
# is capped at "max-interval". (boolean, default: false)
forever = false
```
Long story short, setting `forever = true` and maybe reducing the value of `max-interval` will give you almost live updates even long after the first pipelines have finished. Just be aware that it causes `cistern` to endlessly send requests to your source and CI providers.

For the *g* key you need to either type the name of the reference (or use autocomplete (*TAB* key) to fill the prompt with a reference from the list) and then press *Enter*. *Enter* won't select  a reference from the list, it's just used for finishing the entry. I was afraid this would be confusing but I can't use *Enter* for both autocomplete (like *TAB*) and finishing the entry. Maybe disabling the up/down arrows would encourage typing a reference instead of navigating the list?

**Comment by simplelife2010** [comment]
Works flawlessly, wow!

Something I think about:
Currently cistern monitors commits. It is possible to do multiple pipeline runs on one commit, and with the new feature I will see all of them as they are started. Good.
Now it does not happen often that I run pipeline multiple times on one commit. My default workflow is to push a new commit, and GitLab will trigger the pipeline. This is supported by the 'follow' feature you added recently. Pressing 'f' would show me the pipeline of the newest commit, and the one from the commit before would disappear.

Would it be reasonable to have cistern add the new pipeline to the existing list (also by automated polling) instead of replacing the old one? This would mean that the scope of monitoring would not be a specific commit, but a branch - like master - (or optionally the whole gitlab project as well).

Of course this needed to be conceptually consistent.

Apart from that, I am 98% happy with my use case.

**Comment by nbedos** [comment]
> Would it be reasonable to have cistern add the new pipeline to the existing list (also by automated polling) instead of replacing the old one? This would mean that the scope of monitoring would not be a specific commit, but a branch - like master - (or optionally the whole gitlab project as well).

I tried this approach at the beginning: https://github.com/nbedos/cistern/tree/8a6f554c63e1067018f83cd9871e9e70394adfbc
In the animated demo shown in the README you can see a table. Each line of the table is a commit that can be opened to reveal all the associated CI pipelines, jobs, etc. At this point the scope of the monitoring was probably a branch.

The main problem was that the startup could be quite slow. It takes just a few seconds to list the pipelines of a single commit, but when monitoring a branch you probably want to fetch the pipelines of the last 5 to 10 commits and so now you have to wait 15 seconds or more for `cistern` to populate the table. I was also pretty sure that at some point someone would legitimately ask for a way to go back in the history and that would also be slow and unpleasant to use.

That's why today `cistern` has a more targeted approach and lists only the pipelines of a single commit: there is less data to download than when monitoring a branch or repository, no need to implement some kind of infinite scrolling or pagination that would probably be unpleasant to use. And `cistern` still works with pipelines of 250 jobs like [these](https://gitlab.com/gitlab-org/gitlab/pipelines/).

Another reason for exclusively monitoring commits is that I built `cistern` for the purpose of listing the pipelines associated to what I'm working on at a given time. I almost never care about older pipelines, I just want to see the ones triggered by my last commit. And in the few cases where I do want to see a pipeline unrelated to this commit, jumping between git references with 'g' is good enough.

So to answer your question, no, sorry, any monitoring at a scope broader than a commit will most likely never get implemented.

**Comment by simplelife2010** [comment]
Thanks for the explanation. I agree.

What do you think of a configurable (enabled/disabled) auto-follow feature? If enabled, instead of me pressing f after a new pushed commit, cistern would recognise that by itself after each polling period and start monitoring the new commit.

**Comment by nbedos** [comment]
> What do you think of a configurable (enabled/disabled) auto-follow feature?

I think it's a feature that, for now, doesn't have much value compared to the time it would take to implement. There are other opened issues like #14 and #23 that are far more important and I'd rather work on those. So maybe bring this back up in a month or two? At that point I should be done with these issues and I'll probably be more receptive to discussing changes that fall in the quality-of-life category.

**Comment by nbedos** [comment]
I've merged my changes in `master` and I'm now closing this issue.
